File: internal/database/bookings.go
package database

import (
	"hms-api/models"
	"hms-api/types"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

func (s *service) GetGuests() ([]models.Guest, error) {
	var guests []models.Guest
	result := s.db.Model(&models.Guest{}).Find(&guests)
	return guests, result.Error
}

func (s *service) GetBookings() ([]models.Booking, error) {
	var bookings []models.Booking
	result := s.db.Model(&models.Booking{}).Preload("Guest").Preload("Room").Find(&bookings)
	return bookings, result.Error
}

func (s *service) GetBooking(id uuid.UUID) (*models.Booking, error) {
	var booking models.Booking
	result := s.db.Model(&models.Booking{}).Where("id = ?", id).Preload("Guest").Preload("Room").First(&booking)
	return &booking, result.Error
}

func (s *service) CreateBooking(payload *types.CreateBookingPayload) error {
	tx := s.db.Begin()
	var guest *models.Guest
	guest = &models.Guest{}
	if err := tx.Model(guest).Where("email = ?", payload.Guest.Email).First(guest).Error; err != nil {
		if err == gorm.ErrRecordNotFound { // If no guest with the same email is found, create a new guest
			guest = &models.Guest{
				FullName:    payload.Guest.FullName,
				Email:       payload.Guest.Email,
				PhoneNumber: payload.Guest.PhoneNumber,
				Address:     payload.Guest.Address,
			}
			if err := tx.Create(guest).Error; err != nil {
				tx.Rollback()
				return err
			}
		} else { // If an error occurs while checking for an existing guest, roll back the transaction
			tx.Rollback()
			return err
		}
	}

	// Create the booking
	booking := &models.Booking{
		CheckInDate:   payload.CheckInDate,
		CheckOutDate:  payload.CheckOutDate,
		Duration:      payload.Duration,
		GuestsNumber:  payload.GuestsNumber,
		GuestRequest:  payload.GuestRequest,
		BookingExtras: payload.BookingExtras,
		RoomID:        payload.RoomID,
		GuestID:       guest.ID,
		Price:         payload.Price,
	}

	if payload.BookingStatus != "" {
		booking.BookingStatus = (*models.BookingStatus)(&payload.BookingStatus)
	}

	fo_stat := "Occupied"

	if err := tx.Create(booking).Error; err != nil {
		tx.Rollback()
		return err
	}
	if payload.BookingStatus != "" {
		if err := tx.Model(&models.Room{}).Where("id", booking.RoomID).Updates(map[string]interface{}{"fo_status": (*models.FOStatus)(&fo_stat)}).Error; err != nil {
			tx.Rollback()
			return err
		}
	}

	// Commit the transaction
	if err := tx.Commit().Error; err != nil {
		return err
	}

	return nil
}

func (s *service) DeleteBooking(id uuid.UUID) error {
	result := s.db.Delete(&models.Booking{}, id)
	return result.Error
}

func (s *service) UpdateBooking(id uuid.UUID, payload *types.UpdateBookingPayload) error {
	_updates := types.ParseBookingUpdates(payload)
	result := s.db.Model(&models.Booking{}).Where("id = ?", id).Updates(_updates)
	return result.Error
}

func (s *service) CheckInBooking(id uuid.UUID) error {
	booking, err := s.GetBooking(id)
	if err != nil {
		return err
	}

	room, err := s.GetRoom(booking.RoomID)
	if err != nil {
		return err
	}

	foStat := "Occupied"
	booking_stat := "CheckedIn"
	room.FOStatus = (*models.FOStatus)(&foStat)
	booking.BookingStatus = (*models.BookingStatus)(&booking_stat)

	if err := s.db.Save(&booking).Error; err != nil {
		return err
	}
	result := s.db.Save(&room)

	return result.Error
}

func (s *service) CheckOut(id uuid.UUID) error {
	booking, err := s.GetBooking(id)
	if err != nil {
		return err
	}
	room, err := s.GetRoom(booking.RoomID)
	if err != nil {
		return err
	}
	foStat := "Vacant"
	roomStat := "Clean up"
	booking_stat := "CheckedOut"
	room.FOStatus = (*models.FOStatus)(&foStat)
	room.RoomStatus = (*models.RoomStatus)(&roomStat)
	booking.BookingStatus = (*models.BookingStatus)(&booking_stat)

	if err := s.db.Save(&booking).Error; err != nil {
		return err
	}
	result := s.db.Save(&room)
	return result.Error
}

func (s *service) GetActiveBookings() ([]models.Booking, error) {
	var bookings []models.Booking
	result := s.db.Model(&models.Booking{}).Where("booking_status = ?", "CheckedIn").Preload("Room").Find(&bookings)
	return bookings, result.Error
}

File: frontend/src/api/bookings.ts
import { BookingFormValues } from "@/features/bookings/components/booking-form";
import { api } from "./api";

export const createBooking = async (
  values: BookingFormValues,
  price: number,
  check_out_date: Date,
  booking_status: string
) => {
  const response = await api.post("/bookings", {
    ...values,
    price,
    check_out_date,
    booking_status,
  });
  if (response.status != 201) {
    throw new Error("Could not create booking");
  }

  return response.data;
};

export const getBookings = async () => {
  const response = await api.get("/bookings");
  if (response.status != 200) {
    throw new Error("Failed to fetch bookings");
  }

  return response.data as Booking[];
};

export const getBooking = async (id: string) => {
  const response = await api.get(`/bookings/${id}`);
  if (response.status != 200) {
    throw new Error("Failed to fetch booking");
  }
  return response.data as Booking;
};

export const checkInBooking = async (id: string) => {
  const response = await api.post(`/bookings/check-in/${id}`);
  if (response.status != 200) {
    throw new Error("Failed to check in");
  }
  return response.data;
};

export const checkOutBooking = async (id: string) => {
  const response = await api.post(`/bookings/check-out/${id}`);
  if (response.status != 200) {
    throw new Error("Failed to check out");
  }
  return response.data;
};

export const cancelBooking = async (id: string) => {
  const response = await api.patch(`/bookings/${id}`, {
    booking_status: "Cancelled",
  });
  if (response.status != 200) {
    throw new Error("Failed to cancel booking");
  }
  return response.data;
};

export const deleteBooking = async (id: string) => {
  const response = await api.delete(`/bookings/${id}`);
  if (response.status != 200) {
    throw new Error("Failed to delete booking");
  }
  return response.data;
};
